llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -1030,6 +1036,12 @@ PassBuilder::buildModuleSimplificationPipeline(OptimizationLevel Level,
       Phase != ThinOrFullLTOPhase::ThinLTOPostLink)
     MPM.addPass(SampleProfileProbePass(TM));
 
+  // Instrument function entry and exit before all inlining.
+  if (!isLTOPostLink(Phase)) {
+    MPM.addPass(createModuleToFunctionPassAdaptor(
```

Comment:
actually is it possible to add this into `EarlyFPM` below? it's nice to have fewer module->function adaptors. (e.g. in `EarlyFPM` we run all the function passes on a single function before moving to the next one which is nice for memory locality)

sorry for suggesting this so late, since it'll require updating all the pipeline tests... but first check that none of the other tests fail

Reply:
the check for `Phase != ThinOrFullLTOPhase::FullLTOPostLink` is unnecessary, `buildModuleSimplificationPipeline` isn't called for FullLTO post link

I'm not sure why you're seeing that behavior if we're only running the post-inline instrumenter once in the codegen pipeline

Reply:
if you have a small local repro, you can pass `-Wl,-mllvm,-print-after-all` to the link command to have it print IR after every pass and see where the calls are getting added. or to see changes in the pre-link optimization pipeline, `-mllvm=-print-after-all` (or `-mllvm=-print-changed=quiet`, which doesn't work with the IR parts of the codegen pipeline for *reasons*)

Reply:
> the check for `Phase != ThinOrFullLTOPhase::FullLTOPostLink` is unnecessary, `buildModuleSimplificationPipeline` isn't called for FullLTO post link

Removed. Indeed there is an assertion for this a few lines prior.

> I'm not sure why you're seeing that behavior if we're only running the post-inline instrumenter once in the codegen pipeline

Apparently I made a mistake when building the reproducer. When checking today I saw everything working as expected, i.e. <=1 bare hooks inserted per function in the final DSO. Sorry for the noise.